portal-impl/src/com/liferay/portal/service/impl/RegionLocalServiceImpl.java

```diff
@@ -68,78 +68,109 @@ public void deleteCountryRegions(long countryId) {
 		Region region = regionPersistence.fetchByPrimaryKey(regionId);
 
 		return deleteRegion(region);
 	}
 
 	@Override
-	public Region deleteRegion(Region region) {
+	public Region deleteRegion(Region region) throws PortalException {
+
+		// Region
 
 		regionPersistence.remove(region);
 
-		AddressLocalService.de
-		return null;
+		// Address
+
+		addressLocalService.deleteRegionAddresses(region.getRegionId());
+
+		return region;
 	}
 
 	@Override
 	public Region fetchRegion(long countryId, String regionCode) {
-		return null;
+		return regionPersistence.fetchByC_R(countryId, regionCode);
 	}
 
 	@Override
-	public Region getRegion(long countryId, String regionCode) {
-		return null;
+	public Region getRegion(long countryId, String regionCode)
+		throws PortalException {
+
+		return regionPersistence.findByC_R(countryId, regionCode);
 	}
 
 	@Override
-	public List<Region> getRegions(long countryId, boolean active) {
-		return null;
+	public List<Region> getRegions(long countryId, boolean active)
+		throws PortalException {
+
+		return regionPersistence.findByC_A(countryId, active);
 	}
 
 	@Override
 	public List<Region> getRegions(
 		long countryId, boolean active, int start, int end,
 		OrderByComparator<Region> orderByComparator) {
 
-		return null;
+		return regionPersistence.findByC_A(
+			countryId, active, start, end, orderByComparator);
 	}
 
 	@Override
 	public List<Region> getRegions(
 		long countryId, int start, int end,
 		OrderByComparator<Region> orderByComparator) {
 
-		return null;
+		return regionPersistence.findByCountryId(
+			countryId, start, end, orderByComparator);
 	}
 
 	@Override
-	public List<Region> getRegions(long companyId, String a2, boolean active) {
-		return null;
+	public List<Region> getRegions(long companyId, String a2, boolean active)
+		throws PortalException {
+
+		Country country = countryPersistence.findByC_A2(companyId, a2);
+
+		return regionPersistence.findByC_A(country.getCompanyId(), active);
 	}
 
 	@Override
 	public int getRegionsCount(long countryId) {
-		return null;
+		return regionPersistence.countByCountryId(countryId);
 	}
 
 	@Override
 	public int getRegionsCount(long countryId, boolean active) {
-		return null;
+		return regionPersistence.countByC_A(countryId, active);
 	}
 
 	@Override
-	public Region updateActive(long regionId, boolean active) {
-		return null;
+	public Region updateActive(long regionId, boolean active)
+		throws PortalException {
+
+		Region region = regionPersistence.findByPrimaryKey(regionId);
+
+		region.setActive(active);
+
+		return regionPersistence.update(region);
 	}
 
 	@Override
 	public Region updateRegion(
-		long regionId, boolean active, String name, double position,
-		String regionCode) {
+			long regionId, boolean active, String name, double position,
+			String regionCode)
+		throws PortalException {
+
+		validate(name, regionCode);
+
+		Region region = regionPersistence.findByPrimaryKey(regionId);
```

Comment:
This should go to the top before doing the validate because it's possible that the regionId is invalid. And in that case, we don't need to validate the name and regionCode.